File: cpp/Greedy/55.jump-game.cpp
#include "../std.hpp"
class Solution {
public:
    bool canJump(vector<int> &nums)
    {
        for (int i = 0, reachable = 1; i < nums.size() && reachable > 0; ++i) {
            if (reachable >= nums.size() - i) {
                return true;
            }
            reachable = max(reachable - 1, nums[i]);
        }
        return false;
    }
};

File: TypeScript/Greedy/55.jump-game.ts
function canJump(nums: number[]): boolean {
  let goal = nums.length - 1
  for (let i = goal - 1; i >= 0; --i) {
    if (i + nums[i] >= goal) {
      goal = i
    }
  }
  return goal == 0
}
console.log(canJump([3, 2, 1, 0, 4]))
